persist: don't consolidate unsealed batches

ruchirK opened the pull request:
Previously, we enforced the invariants that data stored in unsealed batches was
- sorted by (time, (key, val))
- consolidated, meaning that each instance of ((key, val), time) occurred exactly
  once.
- diffs were non-zero.

These invariants are potentially valuable in making trace batch construction
more efficient (since records are sorted by time you don't need to scan the entire
unsealed batch), improving the storage space utilization, and holding the data in
a canonical format that makes it trivial to compare if two unsealed batches are
identical. However, we were not using any of these properties currently, and they
were costing us a lot of CPU time in `drain_pending` to enforce, which was directly
contributing to write latency. Also, these properties are harder to enforce in a
columnar representation where its harder to sort and remove records.

This commit removes the three invariants mentioned above, and removes the parts
of drain_pending that enforced them.

mem_indexed_write_drain_sorted
                        time:   [237.57 ms 239.31 ms 241.10 ms]
                        change: [-13.368% -12.430% -11.469%] (p = 0.00 < 0.05)
                        
mem_indexed_write_drain_unsorted
                        time:   [263.83 ms 267.23 ms 270.80 ms]
                        change: [-46.222% -45.339% -44.445%] (p = 0.00 < 0.05)
                        
file_indexed_write_drain_sorted
                        time:   [360.92 ms 362.98 ms 365.07 ms]
                        change: [-10.657% -9.8504% -9.0228%] (p = 0.00 < 0.05)
                        
file_indexed_write_drain_unsorted
                        time:   [386.79 ms 391.11 ms 395.56 ms]
                        change: [-37.397% -36.510% -35.626%] (p = 0.00 < 0.05)

Note that the performance improvement is greater for the tests that write unsorted
data which is to be expected because the stdlib sort probably has some sort of
sortedness-detection that lets it skip a bunch of work when the input is pre-sorted.

ruchirK commented:
will add perf impact from this change once #8893 is in mergeable shape

ruchirK commented:
RFAL! the first commit is just #8893

danhhz commented on src/persist/src/indexed/mod.rs:
```diff
@@ -1367,8 +1352,22 @@ mod tests {
+
         // Now move the data to the trace part of the index, which consolidates
-        // updates at identical ((k, v), t). Since the writes are only consolidated
+        // updates at identical ((k, v), t). Since the writes are unconsolidated
```
nit: remove "within individual unsealed batches"

danhhz commented on src/persist/src/indexed/mod.rs:
```diff
@@ -1367,8 +1352,22 @@ mod tests {
             i.write(vec![(id, updates.clone())], handle)
         })?;
 
+        // Sanity check that the data is all unconsolidated in unsealed and none of
```
nit: it's strange for the test to be actually asserting that we want them unconsolidated. it's worth asserting that it's in unsealed, so I think my issue is just with the phrasing of this comment. let's remove the bit here about them being unconsolidated (and if you like, put something above the assert_eq on line 1358 to the effect of "// Note that unsealed does not consolidate")

danhhz reviewed (approved):
LGTM but I'd like to see throughput numbers in the commit (requiring first that they're added to the benchmark PR)

ruchirK commented:
```
    persist: don't consolidate unsealed batches

    Previously, we enforced the invariants that data stored in unsealed batches was
    - sorted by (time, (key, val))
    - consolidated, meaning that each instance of ((key, val), time) occurred exactly
      once.
    - diffs were non-zero.

    These invariants are potentially valuable in making trace batch construction
    more efficient (since records are sorted by time you don't need to scan the entire
    unsealed batch), improving the storage space utilization, and holding the data in
    a canonical format that makes it trivial to compare if two unsealed batches are
    identical. However, we were not using any of these properties currently, and they
    were costing us a lot of CPU time in `drain_pending` to enforce, which was directly
    contributing to write latency. Also, these properties are harder to enforce in a
    columnar representation where its harder to sort and remove records.

    This commit removes the three invariants mentioned above, and removes the parts
    of drain_pending that enforced them. After doing so, we had to update the golden_test.json
    file, but this did not necessitate a version increment, as we're not changing the
    encoding so much as what we decide to write down, and reading previously consolidated
    data is still supported.

    indexed_write_drain/mem_sorted/39000056
                            time:   [202.54 ms 205.01 ms 207.70 ms]
                            thrpt:  [179.07 MiB/s 181.42 MiB/s 183.64 MiB/s]
    indexed_write_drain/mem_unsorted/39000056
                            time:   [209.66 ms 214.38 ms 219.60 ms]
                            thrpt:  [169.37 MiB/s 173.49 MiB/s 177.40 MiB/s]
    indexed_write_drain/file_sorted/39000056
                            time:   [324.29 ms 329.86 ms 335.31 ms]
                            thrpt:  [110.92 MiB/s 112.76 MiB/s 114.69 MiB/s]
    indexed_write_drain/file_unsorted/39000056
                            time:   [326.70 ms 330.13 ms 334.13 ms]
                            thrpt:  [111.31 MiB/s 112.66 MiB/s 113.85 MiB/s]

    Note that the performance improvement is greater for the tests that write unsorted
    data which is to be expected because the stdlib sort probably has some sort of
    sortedness-detection that lets it skip a bunch of work when the input is pre-sorted.

    Also note that if you compare throughput numbers from when we used abomation to
    bincode, throughput seems down - but this is misleading because that only
    tracks the amount of bytes we write down. Bincode reduces the size of the
    encoded data by approximately 50%, but we're still writing the same underlying
    data, and doing so faster than we were with abom.

````

danhhz commented:
> After doing so, we had to update the golden_test.json file but this did not necessitate a version increment, as we're not changing the encoding so much as what we decide to write down, and reading previously consolidated data is still supported.

Huh I didn't think too hard about this the first time, but I'm curious why we had to update golden_test.json. The previously written data should still be readable, no?

danhhz commented:
For context, in case this is the misunderstanding, the goal of golden_test isn't to represent what we currently write down for some set of operations, but instead to check that what we've previously written down still reads back as what it should. Essentially it's the same as the upgrade tests but at a much lower level of the system

danhhz commented:
> Also note that if you compare throughput numbers from when we used abomation to bincode, throughput seems down - but this is misleading because that only tracks the amount of bytes we write down.

We can (and should) fix this by making them "goodput" numbers. Instead of measuring the actual encoded size, we assign a logical size to each update and use that (I use `key.len() + val.len() + size_of<u64>() + size_of<isize>()`). That will make the throughput numbers comparable across encoding changes

ruchirK commented:
Here's my understanding:

1. golden test keeps around a previous "golden" representation of what data was stored in Indexed after a series of operations
2. golden spins up a fresh indexed and does those same operations on a fresh indexed
3. golden checks that the in-memory contents of (2) are the same as (1)

Since this pr changed unsealed batches to no longer consolidate. that's no longer true. That's the assert that's failing in the code iiuc -- it seems like the assert is on line 130 in golden_test.rs?

ruchirK commented:
Ack on goodput. I'll do that in a separate commit

danhhz commented:
Ah, nah this seems like a bug in golden_test. Ideally it's independent of the actual in-memory state, but instead cares about the (we'll call it) "logical" state of each persisted collection's [since, upper) and its contents. I think the issue is that the bit that slurps the `PersistStreamState` should consolidate the snapshot contents before comparing them

ruchirK commented:
hmmm I actually kind of feel like golden is working as expected. For example, say we changed the sort order in trace / unsealed batches. Then, previously persisted data would not match the current sort order, but if we consolidate before comparing we wouldn't know that? But data in the old sort order might violate some invariants on the new code

danhhz commented:
IMO that's too prescriptive. If the new code depends on a different sort than we've previously written out, then it needs to detect and handle that

ruchirK commented:
Alright I added the goodput measurements in another commit (benchmark data is in the commit message) @danhhz does this look good to merge?

danhhz commented:
I think we need to resolve the golden_test bug before we merge this. It makes me uncomfortable that this PR, which should be able to read back and use previously persisted data (as evidenced by not needing a BlobMeta::CURRENT_VERSION bump), doesn't pass the golden_test without rewriting the golden data.

ruchirK commented:
> I think we need to resolve the golden_test bug before we merge this.

Ack. Sorry I misunderstood earlier - I thought you just meant that there's a bug in golden_test but its benign. Have fixed it locally.

> Also can we get the benchmark change lines in the second commit message plz! These ones:

To clarify, you want the throughput changes from before we started measuring goodput to after we started measuring goodput right? If so, I actually don't think criterion will let me do that because it treats `indexed_write_drain/mem_sorted/foo` and `indexed_write_drain/mem_sorted/bar` where `foo` and `bar` are the badput and goodput measures of size as two distinct, incomparable benchmarks. I tried manually saving a baseline measurement and comparing against that and criterion crashed :)

On a different note - nemesis found a bug. Turns out a huge part of the `Unsealed` implementation relies on the data being sorted by time. Gonna take a bit to fix and probably won't be done tomorrow. Another reminder for me to always run nemesis on changes :/ I'm glad ci caught this at least.

danhhz commented:
> To clarify, you want the throughput changes from before we started measuring goodput to after we started measuring goodput right?

Hmm I'm not sure if this is what I'm saying or not. Concretely: rebase the goodput commit to be first (I can't easily tell if it is already from github's UI, le sigh), then checkout HEAD~1 (the goodput commit) and run benchmarks, then checkout HEAD (the perf commit) and run them again and saving the absolute numbers as well as the change numbers. Does that make any more sense?

ruchirK commented:
Alright this finally hits all of the marks

- fixed golden_test.rs to consolidate before comparing
- fixed unsealed to determine the min/max times by scanning through all updates at append time instead of blindly taking first and last. Getting rid of that second scan is followup.
- Added benchmark diffs to the commit that removes unsealed consolidation. tl;dr its roughly a 15% improvement on the benchmark.

This took _way_ longer than expected today because I accidentally had some stray diff in my golden_test.json, which manifested somehow as the golden test only failing when I changed the code to scan for min and max. Spent forever trying to track that down before realizing the problem was that golden_test.json was itself corrupt.

danhhz commented:
nit: the second commit has an outdated reference to the golden_test.json update that's no longer needed

also, is there a test for whatever regression nemesis caught? if that's easy, it's nice to have direct tests for those kind of things so we don't have to rely on stressing nemesis

danhhz commented on src/persist/src/indexed/unsealed.rs:
```diff
@@ -123,22 +123,30 @@ impl Unsealed {
+
+        let (ts_upper, ts_lower) = {
+            let mut upper = Timestamp::minimum();
+            let mut lower = u64::MAX;
```
nit: this get a little more bulletproof if you store upper and lower as an Option of a 2-tuple as you iterate. and then move the `return Err(... "invalid unsealed batch: trying to write empty batch")` to cover the case where you're done iterating and the Option is still None

ruchirK commented:
Unit test added!

ruchirK commented:
tftr! merging on green
